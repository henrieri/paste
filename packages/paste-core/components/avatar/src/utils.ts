import {
  IconSizeOptions,
  LineHeightOptions,
  FontSizeOptions,
  IconSize,
  LineHeight,
  FontSize,
} from '@twilio-paste/style-props';

export const getCorrespondingLineHeightFromSizeToken = (size: IconSizeOptions): LineHeightOptions =>
  `lineHeight${size.replace('sizeIcon', '')}` as LineHeightOptions;

export const getCorrespondingFontSizeFromSizeToken = (size: IconSizeOptions): FontSizeOptions => {
  switch (size) {
    case 'sizeIcon10':
    default:
      return 'fontSize10';
    case 'sizeIcon20':
      return 'fontSize20';
    case 'sizeIcon30':
      return 'fontSize30';
    case 'sizeIcon40':
    case 'sizeIcon50':
    case 'sizeIcon60':
      return 'fontSize40';
    case 'sizeIcon70':
      return 'fontSize50';
    case 'sizeIcon80':
    case 'sizeIcon90':
      return 'fontSize60';
    case 'sizeIcon100':
      return 'fontSize80';
    case 'sizeIcon110':
      return 'fontSize90';
  }
};

export const getComputedTokenNames = (size: IconSize): {lineHeight: LineHeight; fontSize: FontSize} => {
  if (Array.isArray(size)) {
    return {
      lineHeight: size.map(s => {
        if (s != null) {
          return getCorrespondingLineHeightFromSizeToken(s);
        }
        return null;
      }),
      fontSize: size.map(s => {
        if (s != null) {
          return getCorrespondingFontSizeFromSizeToken(s);
        }
        return null;
      }),
    };
  }
  return {
    lineHeight: getCorrespondingLineHeightFromSizeToken(size as IconSizeOptions),
    fontSize: getCorrespondingFontSizeFromSizeToken(size as IconSizeOptions),
  };
};
