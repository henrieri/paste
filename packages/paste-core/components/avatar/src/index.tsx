import * as React from 'react';
import {Text} from '@twilio-paste/text';
import {Box, safelySpreadBoxProps} from '@twilio-paste/box';
import {IconSize} from '@twilio-paste/style-props';
import {getComputedTokenNames} from './utils';

export interface AvatarProps {
  size: IconSize;
}
export interface AvatarImageProps {
  src: string;
  alt: string;
}

const AvatarContext = React.createContext<AvatarProps>({size: 'sizeIcon10'});

export const AvatarImg: React.FC<AvatarImageProps> = ({src, alt, ...props}) => {
  const {size} = React.useContext(AvatarContext);
  return <Box {...safelySpreadBoxProps(props)} as="img" alt={alt} maxWidth="100%" src={src} size={size} />;
};

const Avatar: React.FC<AvatarProps> = ({children, size = 'sizeIcon70'}) => {
  const computedTokenNames = getComputedTokenNames(size);
  return (
    <AvatarContext.Provider value={{size}}>
      <Box
        as="div"
        backgroundColor="colorBackgroundPrimaryLight"
        borderRadius="borderRadiusCircle"
        overflow="hidden"
        size={size}
      >
        <Text
          as="div"
          fontSize={computedTokenNames.fontSize}
          fontWeight="fontWeightSemibold"
          lineHeight={computedTokenNames.lineHeight}
          textAlign="center"
        >
          {children}
        </Text>
      </Box>
    </AvatarContext.Provider>
  );
};

Avatar.displayName = 'Avatar';

export {Avatar};
