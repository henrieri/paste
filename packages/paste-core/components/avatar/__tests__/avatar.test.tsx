import * as React from 'react';
import {render, screen} from '@testing-library/react';
import {axe} from 'jest-axe';
import {Avatar, AvatarImg} from '../src';
import {
  getCorrespondingLineHeightFromSizeToken,
  getCorrespondingFontSizeFromSizeToken,
  getComputedTokenNames,
} from '../src/utils';

describe('Avatar', () => {
  describe('Utils', () => {
    describe('getCorrespondingLineHeightFromSizeToken', () => {
      it('should return a line height to match icon size', () => {
        expect(getCorrespondingLineHeightFromSizeToken('sizeIcon20')).toEqual('lineHeight20');
      });
    });

    describe('getCorrespondingFontSizeFromSizeToken', () => {
      it('should return a font size to match icon size', () => {
        expect(getCorrespondingFontSizeFromSizeToken('sizeIcon10')).toEqual('fontSize10');
        expect(getCorrespondingFontSizeFromSizeToken('sizeIcon20')).toEqual('fontSize20');
        expect(getCorrespondingFontSizeFromSizeToken('sizeIcon30')).toEqual('fontSize30');
        expect(getCorrespondingFontSizeFromSizeToken('sizeIcon40')).toEqual('fontSize40');
        expect(getCorrespondingFontSizeFromSizeToken('sizeIcon50')).toEqual('fontSize40');
        expect(getCorrespondingFontSizeFromSizeToken('sizeIcon60')).toEqual('fontSize40');
        expect(getCorrespondingFontSizeFromSizeToken('sizeIcon70')).toEqual('fontSize50');
        expect(getCorrespondingFontSizeFromSizeToken('sizeIcon80')).toEqual('fontSize60');
        expect(getCorrespondingFontSizeFromSizeToken('sizeIcon90')).toEqual('fontSize60');
        expect(getCorrespondingFontSizeFromSizeToken('sizeIcon100')).toEqual('fontSize80');
        expect(getCorrespondingFontSizeFromSizeToken('sizeIcon110')).toEqual('fontSize90');
      });
    });

    describe('getComputedTokenNames', () => {
      it('should handle singe size values', () => {
        expect(getComputedTokenNames('sizeIcon50')).toEqual({fontSize: 'fontSize40', lineHeight: 'lineHeight50'});
      });
      it('should handle responsive size values', () => {
        expect(getComputedTokenNames(['sizeIcon50', 'sizeIcon100'])).toEqual({
          fontSize: ['fontSize40', 'fontSize80'],
          lineHeight: ['lineHeight50', 'lineHeight100'],
        });
      });
    });
  });

  describe('intials', () => {
    it('should render responsive css', (): void => {
      const {asFragment} = render(<Avatar size={['sizeIcon10', 'sizeIcon60', 'sizeIcon100']}>ST</Avatar>);
      expect(asFragment()).toMatchSnapshot();
    });
  });
  describe('image', () => {
    it('should render alt and src attributes', (): void => {
      render(
        <Avatar size="sizeIcon10">
          <AvatarImg alt="avatar example" src="/avatars/avatar2.png" />
        </Avatar>
      );
      expect(screen.getByRole('img').getAttribute('src')).toEqual('/avatars/avatar2.png');
      expect(screen.getByRole('img').getAttribute('alt')).toEqual('avatar example');
    });
    it('should render responsive css with an image', (): void => {
      const {asFragment} = render(
        <Avatar size={['sizeIcon30', 'sizeIcon40', 'sizeIcon90']}>
          <AvatarImg alt="avatar example" src="/avatars/avatar2.png" />
        </Avatar>
      );
      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('accessibility', () => {
    it('should have no accessibility violations', async () => {
      const {container} = render(
        <>
          <Avatar size="sizeIcon10">ST</Avatar>
          <Avatar size="sizeIcon10">
            <AvatarImg alt="avatar example" src="/avatars/avatar2.png" />
          </Avatar>
        </>
      );
      const results = await axe(container);
      expect(results).toHaveNoViolations();
    });
  });
});
