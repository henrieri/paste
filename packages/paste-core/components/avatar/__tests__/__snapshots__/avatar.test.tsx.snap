// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Avatar image should render responsive css with an image 1`] = `
<DocumentFragment>
  .emotion-2 {
  box-sizing: border-box;
  background-color: colorBackgroundPrimaryLight;
  border-radius: borderRadiusCircle;
  overflow: hidden;
  width: sizeIcon30;
  height: sizeIcon30;
}

@media screen and (min-width:40em) {
  .emotion-2 {
    width: sizeIcon40;
    height: sizeIcon40;
  }
}

@media screen and (min-width:52em) {
  .emotion-2 {
    width: sizeIcon90;
    height: sizeIcon90;
  }
}

.emotion-1 {
  margin: 0;
  padding: 0;
  font-size: fontSize30;
  font-weight: fontWeightSemibold;
  line-height: lineHeight30;
  text-align: center;
  color: colorText;
}

@media screen and (min-width:40em) {
  .emotion-1 {
    font-size: fontSize40;
    line-height: lineHeight40;
  }
}

@media screen and (min-width:52em) {
  .emotion-1 {
    font-size: fontSize60;
    line-height: lineHeight90;
  }
}

.emotion-0 {
  box-sizing: border-box;
  max-width: 100%;
  width: sizeIcon30;
  height: sizeIcon30;
}

@media screen and (min-width:40em) {
  .emotion-0 {
    width: sizeIcon40;
    height: sizeIcon40;
  }
}

@media screen and (min-width:52em) {
  .emotion-0 {
    width: sizeIcon90;
    height: sizeIcon90;
  }
}

<div
    class="emotion-2"
    overflow="hidden"
  >
    <div
      class="emotion-1"
      color="colorText"
      font-size="fontSize30,fontSize40,fontSize60"
      font-weight="fontWeightSemibold"
    >
      <img
        alt="avatar example"
        class="emotion-0"
        src="/avatars/avatar2.png"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`Avatar intials should render responsive css 1`] = `
<DocumentFragment>
  .emotion-1 {
  box-sizing: border-box;
  background-color: colorBackgroundPrimaryLight;
  border-radius: borderRadiusCircle;
  overflow: hidden;
  width: sizeIcon10;
  height: sizeIcon10;
}

@media screen and (min-width:40em) {
  .emotion-1 {
    width: sizeIcon60;
    height: sizeIcon60;
  }
}

@media screen and (min-width:52em) {
  .emotion-1 {
    width: sizeIcon100;
    height: sizeIcon100;
  }
}

.emotion-0 {
  margin: 0;
  padding: 0;
  font-size: fontSize10;
  font-weight: fontWeightSemibold;
  line-height: lineHeight10;
  text-align: center;
  color: colorText;
}

@media screen and (min-width:40em) {
  .emotion-0 {
    font-size: fontSize40;
    line-height: lineHeight60;
  }
}

@media screen and (min-width:52em) {
  .emotion-0 {
    font-size: fontSize80;
    line-height: lineHeight100;
  }
}

<div
    class="emotion-1"
    overflow="hidden"
  >
    <div
      class="emotion-0"
      color="colorText"
      font-size="fontSize10,fontSize40,fontSize80"
      font-weight="fontWeightSemibold"
    >
      ST
    </div>
  </div>
</DocumentFragment>
`;
